src/peft/tuners/lora/corda.py

```diff
@@ -0,0 +1,347 @@
+
+        # Save cache to disk
+        if cache_file is not None:
+            cache: dict[str, Any] = {}
+            for name, module in target_modules(model, lora_config):
+                cache[f"{name}.corda_method"] = module.corda_method
+                cache[f"{name}.rank"] = module.rank
+                cache[f"{name}.eigens.S_WC"] = module.eigens.S_WC
+                cache[f"{name}.eigens.U_WC"] = module.eigens.U_WC
+                cache[f"{name}.eigens.V_WC"] = module.eigens.V_WC
+
+            os.makedirs(os.path.dirname(cache_file), exist_ok=True)
+            torch.save(cache, cache_file)
+
+
+@torch.no_grad()
+def calib_cov_distribution(
+    model: nn.Module,
+    config: LoraConfig,
+    run_model: Optional[Callable[[], None]],
+    hooked_model: Optional[nn.Module],
+    covariance_file: Optional[str],
+):
+    if covariance_file is not None and os.path.exists(covariance_file) and os.path.getsize(covariance_file) > 0:
+        all_covariance_matrix = torch.load(covariance_file, map_location=get_model_device(model))
+        for name, module in target_modules(model, config):
+            module.covariance_matrix = all_covariance_matrix[name]
+        return
+
+    if run_model is None:
+        raise ValueError("run_model must be specified when covariance file and cache file aren't built.")
+    if hooked_model is None:
+        hooked_model = model
+    hooked_model.eval()
+
+    def hook(module, input, output):
+        input = input[0].detach().squeeze(0).data  ## (context_length = 2048, dim)
+        if not config.corda_config.use_float16_for_covariance:
+            input = input.float()
+        input = input / torch.max(input).abs()
+
+        # check if input is valid
+        if torch.isnan(input).any() or torch.isinf(input).any():
+            raise ValueError("Invalid value found in input, please check your input data.")
+
+        # calculate covariance and check if it's valid
+        covariance = input.t().matmul(input)
+        if torch.isnan(covariance).any() or torch.isinf(covariance).any():
+            raise ValueError(
+                "Invalid value found in covariance. Please file an issue at https://github.com/huggingface/peft/issues."
+            )
+
+        # calculate mean and std
+        mean = input.mean(0)
+        std = input.std(0)
+
+        # add to module
+        module.sample_count += 1
+        module.covariance_matrix += covariance
+        module.mean += mean
+        module.std += std
+
+        # free memory
+        del covariance, input
+
+    for name, module in target_modules(hooked_model, config):
+        module.sample_count = 0
+        module.covariance_matrix = 0
+        module.mean = 0
+        module.std = 0
+        module.register_forward_hook(hook)
+
+    run_model()
+
+    # In some edge cases you might need to hook a model different from the model to add adapters,
+    # this case you would specify `hooked_model` and set it to a different model from `model`.
```

Comment:
Do you have a specific example for this case? If so, it would be worth a ton to mention it in the comment :)

Reply:
This code exists for a very subtle case in our future work related to CorDA, providing the example would introduce many new concepts, and we believe specifying `hooked_model` is very unlikely needed for users, so we won't provide a specific example. If the user actually needs to obtain covariance on a model different from the model to add adapters, just specify `hooked_model` as the model to obtain covariance, we believe this is intuitive and does not need furthur comments. Does this clear up your confusion?